spec/models/upload_spec.rb

```diff
@@ -219,256 +239,152 @@
-                resp = HTTParty.head(
-                  "#{original_url}/#{chunk.sub_path}",
-                  headers: original_auth_header
-                )
-                expect(resp.response.code.to_i).to eq(200)
-              end
-            }
+          expect(mocked_storage_provider).to receive(:purge)
+            .with(subject)
+            .and_raise(unexpected_exception)
+
+          expect {
+            expect {
+              subject.purge_storage
+            }.to change{Chunk.count}.by(-original_chunks_count)
+          }.to raise_error(unexpected_exception)
+
+          subject.reload
+          expect(subject.purged_on).to be_nil
+          expect(subject.purged_on).to be_nil
+          expect(subject.chunks.count).to eq 0
+        }
+      end
+
+      context 'no StorageProviderException' do
+        it 'should purge successfully' do
+          subject.chunks.each do |chunk|
+            expect(chunk).to receive(:purge_storage)
           end
-        end
 
-        context 'no StorageProviderException' do
-          it {
-            original_chunks = subject.chunks.all
-            resp = HTTParty.head(
-              "#{subject.storage_provider.storage_url}/#{subject.sub_path}",
-              headers: subject.storage_provider.auth_header
-            )
-            expect(resp.response.code.to_i).to eq(200)
-            subject.chunks.each do |chunk|
-              resp = HTTParty.head(
-                "#{subject.storage_provider.storage_url}/#{chunk.sub_path}",
-                headers: subject.storage_provider.auth_header
-              )
-              expect(resp.response.code.to_i).to eq(200)
-            end
-            purge_time = DateTime.now
+          expect(mocked_storage_provider).to receive(:purge)
+            .with(subject)
+
+          purge_time = DateTime.now
+          expect {
             expect {
               subject.purge_storage
-            }.to change{Chunk.count}.by(-original_chunks.length)
-            subject.reload
-            expect(subject.purged_on).not_to be_nil
-            expect(subject.purged_on).to be >= purge_time
-            expect(subject.chunks.count).to eq(0)
-
-            resp = HTTParty.head(
-              "#{subject.storage_provider.storage_url}/#{subject.sub_path}",
-              headers: subject.storage_provider.auth_header
-            )
-            expect(resp.response.code.to_i).to eq(404)
-            original_chunks.each do |chunk|
-              resp = HTTParty.head(
-                "#{subject.storage_provider.storage_url}/#{chunk.sub_path}",
-                headers: subject.storage_provider.auth_header
-              )
-              expect(resp.response.code.to_i).to eq(404)
-            end
-          }
+            }.to change{Chunk.count}.by(-original_chunks_count)
+          }.not_to raise_error
+
+          subject.reload
+          expect(subject.purged_on).not_to be_nil
+          expect(subject.purged_on).to be >= purge_time
```

Comment:
If you used the TimeHelpers to freeze time, the above two lines could be replaced with `expect(subject.purged_on).to eq purge_time`

To use the TimeHelpers, add the following inside `Rspec.describe Upload..` block:
```
include ActiveSupport::Testing::TimeHelpers
```
And then just add the following to whatever context needs time frozen:
```
  around(:each) do |example|
    travel_to(Time.now) do #freeze_time
      example.run
    end
  end
```